### PR TITLE
Add pre-commit and fix findings

### PR DESCRIPTION
**Pull request by dbast**
### Description

Inspired by the .pre-commit-config.yaml in conda/conda-build/conda-pack where similar PRs have been done in the past.

The .pre-commit-config.yaml has been stripped down to mainly only contain the flake8 and shellcheck hooks for keeping the number of changed lines as small as possible. More hooks (isort, pyupgrade) can be incrementally enabled via more PRs. Similar PRs have been e.g. merged to conda without causing too much trouble with existing open PRs.

Getting some linting running automatically is of great help when reviewing PRs. This once requires all current finding being fixed, which is done as part of this PR.

This also finds some not existing code in conda_interface.py where the function `mkdir_p_sudo_safe(cache_dir)` does not exist.

### Checklist - did you ...

- [ ] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [ ] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

**Comment by jaimergp** [comment]
I am aching to do this with the codebase but it will create conflicts on opened PRs so I am inclined to merge as much as possible before we merge this one to minimize the amount of rebasing hell we need :D

**Comment by dbast** [comment]
@jaimergp going to strip down the pre-commit config to reduce the amount of changes... more pre-commit hooks in be incrementally added via more PRs.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      - id: check-ast
+      - id: fix-byte-order-marker
+      - id: check-case-conflict
+      - id: check-merge-conflict
+      - id: debug-statements
+      - id: detect-private-key
+      - id: mixed-line-ending
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+      - id: check-merge-conflict
```
Comment (jaimergp) [inline .pre-commit-config.yaml:16]: duplicate! :D
```suggestion
```

**Comment by jaimergp** [comment]
Ok, that makes more sense. Maybe using `blacken` in the next PR is a good step too.

**Comment by jezdez** [comment]
@dbast I've just added the precommit.ci app to this repo. https://results.pre-commit.ci/repo/github/51571636

Badge: [pre-commit.ci status](https://results.pre-commit.ci/latest/github/conda/constructor/main)

```
[![pre-commit.ci status](https://results.pre-commit.ci/badge/github/conda/constructor/main.svg)](https://results.pre-commit.ci/latest/github/conda/constructor/main)
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -22,6 +22,14 @@ defaults:
...
     shell: bash
 jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # tag=v3
+      - uses: actions/setup-python@2c3dd9e7e29afd70cc0950079bde6c979d1f69f9 # v4
+        with:
+          python-version: '3.9'
+      - uses: pre-commit/action@646c83fcd040023954eafda54b4db0192ce70507 # tag=v3.0.0
```
Comment (jezdez) [inline .github/workflows/main.yml:32]:
```suggestion
```

Not needed with pre-commit.ci anymore.

**Review by jezdez — CHANGES_REQUESTED** [review]

**Comment by dbast** [comment]
@jezdez Any idea how to extend the pre-commit,ci to be able to execute hooks that need additional tools? like the shellcheck hook which requires Docker (or in other cases hooks that require golang)? That is pretty easy with the pre-commit action, as there are anyways more tools pre-installed on the action runners and also extending those can be done via an action step running before the pre-commit step.

This repo has many shellscripts as part of the code and as part of the examples and as shellcheck pretty good detects all kind of issues like unquoted variables, posix compatibility, missing error handling etc.

This PR only checks the *example* shell scripts via shellcheck. The shell scripts in code are templated and require rendering before linting them via shell check; that is done in this PR https://github.com/conda/constructor/pull/600

**Comment by dbast** [comment]
@jezdez we can also invite the (public) renovate bot which to keeps the .pre-commit-config.yml hooks versions updated via PRs it proposes... renovate can also update arbitrary pinning versions, if the right regex pattern are added. That would e.g. help to pin the micromamba version in https://github.com/conda/constructor/pull/605 to have stable CI and have it updated by renovate.

**Comment by jaimergp** [comment]
What about this alternative? https://pypi.org/project/shellcheck-py/

**Comment by dbast** [comment]
@jaimergp hah. good finding. trying that.

**Review by jezdez — APPROVED** [review]

**Pull request merged**